presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
+            return createZeroStats(superset);
+        }
+
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        result.setOutputRowCount(outputRowCount);
+
+        superset.getSymbolsWithKnownStatistics().forEach(symbol -> {
+            SymbolStatsEstimate supersetSymbolStats = superset.getSymbolStatistics(symbol);
+            SymbolStatsEstimate subsetSymbolStats = subset.getSymbolStatistics(symbol);
+
+            SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
+
+            // for simplicity keep the average row size the same as in the input
+            // in most cases the average row size doesn't change after applying filters
+            newSymbolStats.setAverageRowSize(supersetSymbolStats.getAverageRowSize());
+
+            // nullsCount
+            double supersetNullsCount = supersetSymbolStats.getNullsFraction() * supersetRowCount;
+            double subsetNullsCount = subsetSymbolStats.getNullsFraction() * subsetRowCount;
+            double newNullsCount = supersetNullsCount - subsetNullsCount;
+            verify(isNaN(newNullsCount) || newNullsCount >= 0, "newNullsCount must greater than or equal to zero: %s", newNullsCount);
+            newSymbolStats.setNullsFraction(min(newNullsCount, outputRowCount) / outputRowCount);
+
+            // distinctValuesCount
+            double supersetDistinctValues = supersetSymbolStats.getDistinctValuesCount();
+            double subsetDistinctValues = subsetSymbolStats.getDistinctValuesCount();
+            double newDistinctValuesCount;
+            if (isNaN(supersetDistinctValues) || isNaN(subsetDistinctValues)) {
+                newDistinctValuesCount = NaN;
+            }
+            else if (supersetDistinctValues == 0) {
+                newDistinctValuesCount = 0;
+            }
+            else if (subsetDistinctValues == 0) {
+                newDistinctValuesCount = supersetDistinctValues;
+            }
+            else {
+                double supersetNonNullsCount = supersetRowCount - supersetNullsCount;
+                double subsetNonNullsCount = subsetRowCount - subsetNullsCount;
+                double supersetValuesPerDistinctValue = supersetNonNullsCount / supersetDistinctValues;
+                double subsetValuesPerDistinctValue = subsetNonNullsCount / subsetDistinctValues;
+                if (supersetValuesPerDistinctValue <= subsetValuesPerDistinctValue) {
+                    newDistinctValuesCount = supersetDistinctValues - subsetDistinctValues;
+                }
+                else {
+                    newDistinctValuesCount = supersetDistinctValues;
+                }
+            }
+            verify(isNaN(newDistinctValuesCount) || newDistinctValuesCount >= 0, "newDistinctValuesCount must be greater or equal than zero: %s", newDistinctValuesCount);
+            newSymbolStats.setDistinctValuesCount(newDistinctValuesCount);
+
+            // range
```

Comment:
just preserve superset range here (see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R95). When trimming ranges we should account for NDVs density, but that's a whole different story out of scope for this PR

Reply:
Why should we preserve the input range if we can do better for negation? There's even an explicit test that covers that case: https://github.com/prestodb/presto/commit/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-9dbc32793a40a33af0bc1e8018d6b346R302

Reply:
Few reasons:
1. If the number of row per distinct value ("density check") in that particular range (that we subtract) is not enough, then `superset` range should be preserved.
2. If the `superset` has `N` distinct values in intersected range and `subset` has `M` distinct values in intersected range, then `superset` range should be preserved.

Reply:
I don't understand that. If full range for some column `A` is `[0, 5]`. Negated range is `[3,5]`. That means that all the values filtered by negated predicate are within `[3,5]`. That means that the only values left are within `[0, 3]` range. Why should it be conditional, and be based on the density check?

Reply:
Let me show an example:

> If the number of row per distinct value ("density check") in that particular range (intersection that we subtract) is not enough, then superset range should be preserved.

```
superset=[1,1,2,2,3,3,4,4], range(superset)=[1-4]
subset=[3,4], range(subset)=[3-4]
superset - subset=[1,1,2,2,3,4], range(superset - subset)=[1-4]
```

> If the superset has N distinct values in intersected range and subset has M distinct values in intersected range where N > M, then superset range should be preserved.

```
superset = [1,1,2,2,3.99999,3.99999,4,4], range(superset)=[1-4]
subset=[2,2,4,4], range(subset)=[2-3.99999]
superset - subset=[1,1,3.99999,3.99999], range(superset - subset)=[1-3.99999]
```

If you want to have better range negation, you should pushdown NOT to leaf predicates.

Reply:
Yes, you are right. I keep forgetting that range may be filtered by some other predicate on some other column. That actually makes sense. So it feels like we should remove the range negation altogether.

> If you want to have better range negation, you should pushdown NOT to leaf predicates.

What do you mean by that?

Reply:
> What do you mean by that?

I you have expression `NOT(a < 0 OR a > 10)` you will get better range estimate if you rewrite the expression to: `a >= 0 AND a <= 10`. This is because then we avoid `OR` whatsoever.

> So it feels like we should remove the range negation altogether.

I will take a shot at that (to account issues that I've mentioned) when this PR lands if you don't mind.